kong/plugins/bot-detection/handler.lua

```diff
@@ -19,10 +44,79 @@ local function get_user_agent()
   return user_agent
 end
 
+local function examine_agent(user_agent, conf)
+  user_agent = strip(user_agent)
+
+  if conf.whitelist then
+    for _, rule in ipairs(conf.whitelist) do
+      if re_match(user_agent, rule) then
+        return MATCH_WHITELIST
+      end
+    end
+  end
+
+  if conf.blacklist then
+    for _, rule in ipairs(conf.blacklist) do
+      if re_match(user_agent, rule) then
+        return MATCH_BLACKLIST
+      end
+    end
+  end
+
+  for _, rule in ipairs(rules.bots) do
+    if re_match(user_agent, rule) then
+      return MATCH_BOT
+    end
+  end
+
+  return MATCH_EMPTY
+end
+
 function BotDetectionHandler:new()
   BotDetectionHandler.super.new(self, "bot-detection")
 end
 
+function BotDetectionHandler:init_worker()
+  local singletons    = require "kong.singletons"
+  local worker_events = singletons.worker_events
+  local dao_factory   = singletons.dao
+
+  -- load our existing plugins to create our cache spaces
+  local plugins, err = dao_factory.plugins:find_all({
+    name = "bot-detection",
+  })
+  if err then
+    ngx.log(ngx.ERR, "err in fetching plugins: ", err)
+  end
+
+  for i = 1, #plugins do
+    ua_cache[plugins[i].api_id] = {}
+  end
+
+  -- catch updates and tell each worker to adjust their cache
+  -- accordingly. for create/updates we create/purge the cache
+  -- by simply creating a new lua table; for deletions, as we no
+  -- longer need this cache data, we just remove it
+  worker_events.register(function(data)
+    if data.entity.name == "bot-detection" then
+      worker_events.post("bot-detection-invalidate", data.operation,
+                        data.entity)
+    end
+  end, "crud", "plugins")
+
+  worker_events.register(function(entity)
+    ua_cache[entity.api_id] = {}
```

Comment:
One thing we are very concerned with are such tables that can potentially grow indefinitely.

A past example was the initial Lua-land cache implementation that was developed, which was identically built (a simple Lua table). With the high cardinality of the keys that were stored in that table, and given enough time, we could potentially fill-up the memory of our Lua VM if the workers were not rotated with new, fresher ones. We promptly switched to a proper implementation using lua-resty-lrucache.

I feel like we are facing a similar situation here, with high cardinality keys (User-Agents), and a very low cleanup ratio (CRUD operation on the plugin).

It would probably be safer to use lua-resty-lrucache here, with a hard-coded, reasonable size.